akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamPublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import akka.actor._
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object OutputStreamPublisher {
+  def props(bytesReadPromise: Promise[Long], bufSize: Int) = {
+    require(bufSize > 0, "buffer size must be > 0")
+    Props(classOf[OutputStreamPublisher], bytesReadPromise, bufSize)
+      .withDeploy(Deploy.local)
+  }
+
+  trait OutputStreamPublisherInMessage
+
+  case class Write(bs: ByteString) extends OutputStreamPublisherInMessage
+
+  case class Flush() extends OutputStreamPublisherInMessage
+
+  case class Close() extends OutputStreamPublisherInMessage
+
+  trait OutputStreamPublisherOutMessage
+
+  case class Ok() extends OutputStreamPublisherOutMessage
+  case class Failure(msg: String) extends OutputStreamPublisherOutMessage
+
+}
+
+/** INTERNAL API */
+private[akka] class OutputStreamPublisher(bytesReadPromise: Promise[Long], bufSize: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+  var bytesGotTotal = 0L
+
+  var notifyWhenSend: Option[ActorRef] = None
+  var notifyWhenSendAll: Option[ActorRef] = None
+
+  def receive = {
+    //Downstream messages
+    case ActorPublisherMessage.Request(elements) ⇒
+      signalDownstream()
+    case ActorPublisherMessage.Cancel ⇒
+      trySignalOnSend(signalErrorUpstream(_, "Stream is cancelled"))
+      context.stop(self)
+
+    //Upstream messages from OutputStreamAdapter
+    case OutputStreamPublisher.Write(bs) ⇒ onWrite(bs) //TODO what if it's been cancelled? need test
+    case OutputStreamPublisher.Flush()   ⇒ flush()
+    case OutputStreamPublisher.Close()   ⇒ flush(); onCompleteThenStop()
+  }
+
+  def signalDownstream(): Unit =
+    if (isActive) {
+      while (totalDemand > 0 && availableChunks.nonEmpty) {
```

Comment:
Here that's OK, since availableChunks won't be huge. In general hint that we had to optimise those to send in "smaller bursts" sometimes. But here it's not a problem I think, keep as is.